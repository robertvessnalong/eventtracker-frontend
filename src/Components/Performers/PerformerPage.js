import { useState, useEffect, useContext } from 'react';
import { Container, Row, Col, Spinner, Alert } from 'react-bootstrap';
import { Link } from 'react-router-dom';
import UserContext from '../../Context/UserContext';
import EventFinderApi from '../../API/api';
import EventItem from '../Events/EventItem';
import { AiFillHeart, AiOutlineHeart } from 'react-icons/ai';
import uuid from 'react-uuid';

const PerformerPage = () => {
  const { user, setUser, loggedIn } = useContext(UserContext);
  const [performer, setPerformer] = useState([]);
  const [performerEvents, setPerformerEvents] = useState([]);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState(false);
  const [errorMsg, setErrorMsg] = useState([]);
  const [hasEvents, setHasEvents] = useState(false);

  useEffect(() => {
    const getPerformer = async () => {
      const performerId = window.location.href.split('/').pop();
      const performer = await EventFinderApi.getPerformer({ performerId });
      const events = await EventFinderApi.getEvents({ search: performer.name });
      if (performer.error) {
        setError(true);
        setErrorMsg(performer.error);
      } else {
        if (events.length > 0) {
          setHasEvents(false);
          setPerformerEvents(events);
        } else {
          setHasEvents(true);
        }
      }
      setPerformer(performer);
      setLoading(false);
    };
    getPerformer();
  }, []);

  if (loading) {
    return (
      <Row>
        <Col style={{ textAlign: 'center' }}>
          <Spinner animation='border' role='status'>
            <span className='visually-hidden'>Loading...</span>
          </Spinner>
        </Col>
      </Row>
    );
  }

  // Handle Favorties
  const handleFavorites = async (e, performer) => {
    const eventData = {
      type: 'performer',
      performerId: performer.id,
      performerName: performer.name,
      performerImage: performer.image,
      performerUrl: performer.url,
      uuid: user['uuid'],
    };
    const res = e.target.classList.contains('favorited')
      ? await EventFinderApi.removefavorite(eventData)
      : await EventFinderApi.favoriteEvent(eventData);
    if (res.error) {
      console.error(res.error);
    } else {
      const user = JSON.parse(localStorage.getItem('user'));
      const userData = await EventFinderApi.getUserInfo(user.user, user.token);
      setUser(userData);
    }
  };

  return (
    <section className='PerformerPage'>
      <header>
        <h2>{error ? errorMsg : performer.name}</h2>
        {!error && (
          <>
            {loggedIn && (
              <>
                {EventFinderApi.favoriteCheck(user.favorites, performer) ? (
                  <div
                    className='favorite-btn favorited'
                    onClick={(e) => handleFavorites(e, performer)}
                  >
                    <AiFillHeart />
                  </div>
                ) : (
                  <div
                    className='favorite-btn'
                    onClick={(e) => handleFavorites(e, performer)}
                  >
                    <AiOutlineHeart />
                  </div>
                )}
              </>
            )}
          </>
        )}
        <hr></hr>
      </header>
      <Container className='Performer-Events'>
        {hasEvents && (
          <Alert className='mt-4' variant='warning'>
            Sorry, no events found! Try Again
          </Alert>
        )}
        <Row xs={1} xl={2}>
          {performerEvents.map((event) => (
            <Col key={uuid()}>
              <EventItem event={event} favorites={user.favorites} />
            </Col>
          ))}
        </Row>
      </Container>
      <footer>
        <Link className='blue-btn' to='/performers'>
          Go Back
        </Link>
      </footer>
    </section>
  );
};

export default PerformerPage;
