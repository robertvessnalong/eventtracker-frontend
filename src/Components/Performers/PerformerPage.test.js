import {
  render,
  screen,
  fireEvent,
  waitFor,
  waitForElementToBeRemoved,
  getByPlaceholderText,
  getByLabelText,
} from '@testing-library/react';
import '@testing-library/jest-dom';
import PerformerPage from './PerformerPage';
import TestUser from '../../Testing/TestUser';
import UserContext from '../../Context/UserContext';
import { BrowserRouter } from 'react-router-dom';

describe('Events Component', () => {
  // Smoke Test
  test('renders without crashing', async () => {
    render(
      <BrowserRouter>
        <UserContext.Provider value={{ user: TestUser, loggedIn: true }}>
          <PerformerPage />
        </UserContext.Provider>
      </BrowserRouter>
    );
    await waitFor(() => {
      const headerSpan = screen.getByText(/Billy Joel/i);
      expect(headerSpan).toBeInTheDocument();
    });
  });

  // Snapshot
  test('matches snapshot', async () => {
    const { asFragment } = render(
      <BrowserRouter>
        <UserContext.Provider value={{ user: TestUser }}>
          <PerformerPage />
        </UserContext.Provider>
      </BrowserRouter>
    );
    await waitForElementToBeRemoved(screen.getByText('Loading...'));
    expect(asFragment()).toMatchSnapshot();
  });
});
