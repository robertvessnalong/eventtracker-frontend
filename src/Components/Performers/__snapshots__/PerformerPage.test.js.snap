// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Events Component matches snapshot 1`] = `
<DocumentFragment>
  <section
    class="PerformerPage"
  >
    <header>
      <h2>
        Billy Joel
      </h2>
      <hr />
    </header>
    <div
      class="Performer-Events container"
    >
      <div
        class="row row-cols-xl-2 row-cols-1"
      >
        <div
          class="col"
        >
          <div
            class="mt-4 mb-4 p-0 card"
          >
            <h6
              class="card-header"
            >
              <div
                class="row"
              >
                <div
                  class="d-flex align-items-center col-10"
                >
                  <small>
                    The Great Jack OLantern Blaze - Long Island
                  </small>
                </div>
                <div
                  class="col-2"
                />
              </div>
            </h6>
            <div
              class="card-body"
            >
              <div
                class="row"
              >
                <div
                  class="text-start col-md-6 col-12"
                >
                  <div
                    class="card-time"
                  >
                    <b>
                      Sun Oct 24 2021 - 7 AM
                    </b>
                  </div>
                  <address>
                    1303 Round Swamp Road
                    <br />
                    Old Bethpage, NY 11804
                  </address>
                </div>
                <div
                  class="d-flex align-items-center justify-content-end col-md-6 col-12"
                >
                  <a
                    href="https://seatgeek.com/the-great-jack-olantern-blaze-long-island-tickets/family/2021-10-24-3-30-am/5491086"
                    rel="noreferrer"
                    target="_blank"
                  >
                    <button
                      class="blue-btn btn btn-primary"
                      type="button"
                    >
                      Buy Tickets
                    </button>
                  </a>
                </div>
              </div>
            </div>
            <div
              class="text-muted text-start card-footer"
            >
              <button>
                Comment
              </button>
              <div
                class="Comment col-12"
              >
                <div
                  class="container"
                >
                  <h6>
                    Comment On Event
                  </h6>
                  <form
                    class=""
                  >
                    <textarea
                      class="form-control"
                      name="comment"
                      placeholder="Leave a comment here"
                      style="height: 150px;"
                    />
                    <button
                      type="Submit"
                    >
                      Submit
                    </button>
                  </form>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <footer>
      <a
        class="blue-btn"
        href="/performers"
      >
        Go Back
      </a>
    </footer>
  </section>
</DocumentFragment>
`;
